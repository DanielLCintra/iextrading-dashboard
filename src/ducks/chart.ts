import { AnyAction, Dispatch } from 'redux';
import Api from '../utils/Api';

export const DUCK_NAME = 'chart';

// Interfaces

export interface ICandlestick {
  date: string,
  open: number,
  high: number,
  low: number,
  close: number,
  label: string,
}

export interface IChartState {
  data: ICandlestick[] | null,
  error: Error | null,
  loading: boolean,
}

// Action types

export const LOAD_CHART_STARTED = `${DUCK_NAME}/LOAD_CHART_STARTED`;
interface ILoadChartStarted {
  symbol: string,
  type: typeof LOAD_CHART_STARTED,
}

export const LOAD_CHART_SUCCEED = `${DUCK_NAME}/LOAD_CHART_SUCCEED`;
interface ILoadChartSucceed {
  symbol: string,
  type: typeof LOAD_CHART_SUCCEED,
  data: ICandlestick[],
}

export const LOAD_CHART_FAILED = `${DUCK_NAME}/LOAD_CHART_FAILED`;
interface ILoadChartFailed {
  symbol: string,
  type: typeof LOAD_CHART_FAILED,
  error: Error,
}

// Action creators

export const loadChartStarted = (symbol:string):ILoadChartStarted =>
    ({ symbol, type: LOAD_CHART_STARTED });

export const loadChartSucceed = (symbol:string, data:ICandlestick[]):ILoadChartSucceed =>
    ({ symbol, type: LOAD_CHART_SUCCEED, data });

export const loadChartFailed = (symbol:string, error:Error):ILoadChartFailed =>
  ({ symbol, type: LOAD_CHART_FAILED, error });

// Thunks

export const loadChart = (symbol:string) =>
  async (dispatch:Dispatch) => {
    dispatch(loadChartStarted(symbol));
    try {
      const results = await Api.instance.getChart(symbol);
      dispatch(loadChartSucceed(symbol, results.data.data));
      return Promise.resolve(results);
    } catch (error) {
      dispatch(loadChartFailed(symbol, error));
      return Promise.reject(error);
    }
  };

// Reducer

export const INITIAL_STATE:Map<string, IChartState> = new Map();

export const INITIAL_CHART_STATE:IChartState = {
  data: null,
  error: null,
  loading: false,
};

const reducer = (state = INITIAL_STATE, action:AnyAction) => {
  const company = state[action.symbol] || INITIAL_CHART_STATE;

  switch (action.type) {
  case LOAD_CHART_STARTED:
    return new Map(state)
      .set(action.symbol, {
        ...company,
        loading: true,
      });
  case LOAD_CHART_SUCCEED:
    return new Map(state)
    .set(action.symbol, {
      data: action.data,
      error: null,
      loading: false,
    });
  case LOAD_CHART_FAILED:
    return new Map(state)
    .set(action.symbol, {
      data: null,
      error: action.error,
      loading: false,
    });;
  default: return state;
  }
};

export default reducer;
