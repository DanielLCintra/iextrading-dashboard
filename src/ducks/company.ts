import { AnyAction, Dispatch } from 'redux';
import Api from '../utils/Api';

export const DUCK_NAME = 'company';

// Interfaces

export interface ICompany {
  symbol: string,
  companyName: string,
  industry: string,
  website: string,
  description: string,
  CEO: string,
  sector: string,
  tags: string[],
}

export interface ICompanyState {
  data: ICompany | null,
  error: Error | null,
  loading: boolean,
}

// Action types

export const LOAD_COMPANY_STARTED = `${DUCK_NAME}/LOAD_COMPANY_STARTED`;
interface ILoadCompanyStarted {
  symbol: string,
  type: typeof LOAD_COMPANY_STARTED,
}

export const LOAD_COMPANY_SUCCEED = `${DUCK_NAME}/LOAD_COMPANY_SUCCEED`;
interface ILoadCompanySucceed {
  symbol: string,
  type: typeof LOAD_COMPANY_SUCCEED,
  data: ICompany[],
}

export const LOAD_COMPANY_FAILED = `${DUCK_NAME}/LOAD_COMPANY_FAILED`;
interface ILoadCompanyFailed {
  symbol: string,
  type: typeof LOAD_COMPANY_FAILED,
  error: Error,
}

// Action creators

export const loadCompanyStarted = (symbol:string):ILoadCompanyStarted =>
    ({ symbol, type: LOAD_COMPANY_STARTED });

export const loadCompanySucceed = (symbol:string, data:ICompany[]):ILoadCompanySucceed =>
    ({ symbol, type: LOAD_COMPANY_SUCCEED, data });

export const loadCompanyFailed = (symbol:string, error:Error):ILoadCompanyFailed =>
  ({ symbol, type: LOAD_COMPANY_FAILED, error });

// Thunks

export const loadCompany = (symbol:string) =>
  async (dispatch:Dispatch) => {
    dispatch(loadCompanyStarted(symbol));
    try {
      const results = await Api.instance.getCompany(symbol);
      dispatch(loadCompanySucceed(symbol, results.data));
      return Promise.resolve(results);
    } catch (error) {
      dispatch(loadCompanyFailed(symbol, error));
      return Promise.reject(error);
    }
  };

// Reducer

export const INITIAL_STATE:Map<string, ICompanyState> = new Map();

const reducer = (state = INITIAL_STATE, action:AnyAction) => {
  switch (action.type) {
  case LOAD_COMPANY_STARTED:
    return new Map(state)
      .set(action.symbol, {
        data: null,
        error: null,
        loading: true,
      });
  case LOAD_COMPANY_SUCCEED:
    return new Map(state)
    .set(action.symbol, {
      data: action.data,
      error: null,
      loading: false,
    });
  case LOAD_COMPANY_FAILED:
    return new Map(state)
    .set(action.symbol, {
      data: null,
      error: action.error,
      loading: false,
    });;
  default: return state;
  }
};

export default reducer;
