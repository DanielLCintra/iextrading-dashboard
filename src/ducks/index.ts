import { combineReducers } from 'redux';

import stocksReducer, {
  DUCK_NAME as STOCKS_DUCK_NAME
} from './stocks';

import symbolsReducer, {
  DUCK_NAME as SYMBOLS_DUCK_NAME
} from './symbols';

import companyReducer, {
  DUCK_NAME as COMPANY_DUCK_NAME
} from './company';

import chartReducer, {
  DUCK_NAME as CHART_DUCK_NAME
} from './chart';

export default combineReducers({
  [STOCKS_DUCK_NAME]: stocksReducer,
  [SYMBOLS_DUCK_NAME]: symbolsReducer,
  [COMPANY_DUCK_NAME]: companyReducer,
  [CHART_DUCK_NAME]: chartReducer,
});
