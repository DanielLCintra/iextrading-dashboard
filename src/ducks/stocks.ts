import { AnyAction, Dispatch } from 'redux';
import Api from '../utils/Api';

export const DUCK_NAME = 'stocks';

// Interfaces

export interface IStock {
  symbol: string,
  companyName: string,
  sector: string,
  latestPrice: number,
  ytdChange: number,
}

export interface IStocksState {
  data: IStock[] | null,
  error: Error | null,
  loading: boolean,
}

// Action types

export const LOAD_STOCKS_STARTED = `${DUCK_NAME}/LOAD_STOCKS_STARTED`;
interface ILoadStocksStarted {
  type: typeof LOAD_STOCKS_STARTED,
}

export const LOAD_STOCKS_SUCCEED = `${DUCK_NAME}/LOAD_STOCKS_SUCCEED`;
interface ILoadStocksSucceed {
  type: typeof LOAD_STOCKS_SUCCEED,
  data: IStock[],
}

export const LOAD_STOCKS_FAILED = `${DUCK_NAME}/LOAD_STOCKS_FAILED`;
interface ILoadStocksFailed {
  type: typeof LOAD_STOCKS_FAILED,
  error: Error,
}

// Action creators

export const loadStocksStarted = ():ILoadStocksStarted =>
  ({ type: LOAD_STOCKS_STARTED });

export const loadStocksSucceed = (data:IStock[]):ILoadStocksSucceed =>
  ({ type: LOAD_STOCKS_SUCCEED, data });

export const loadStocksFailed = (error:Error):ILoadStocksFailed =>
  ({ type: LOAD_STOCKS_FAILED, error });

// Thunks

export const loadStocks = () =>
  async (dispatch:Dispatch) => {
    dispatch(loadStocksStarted());
    try {
      const results = await Api.instance.getStocks();
      dispatch(loadStocksSucceed(results.data));
      return Promise.resolve(results);
    } catch (error) {
      dispatch(loadStocksFailed(error));
      return Promise.reject(error);
    }
  };

// Reducer

export const INITIAL_STATE:IStocksState = {
  data: null,
  error: null,
  loading: false,
};

const reducer = (state = INITIAL_STATE, action:AnyAction) => {
  switch (action.type) {
  case LOAD_STOCKS_STARTED:
    return {
      ...state,
      loading: true,
    };
  case LOAD_STOCKS_SUCCEED:
    return {
      ...state,
      data: [
        ...(state.data || []),
        ...action.data,
      ],
      loading: false,
    };
  case LOAD_STOCKS_FAILED:
    return {
      ...state,
      error: action.error,
      loading: false,
    };
  default: return state;
  }
};

export default reducer;
