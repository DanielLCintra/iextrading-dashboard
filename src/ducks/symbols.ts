import { AnyAction, Dispatch } from 'redux';
import Api from '../utils/Api';

export const DUCK_NAME = 'symbols';

// Interfaces

export interface ISymbol {
  symbol: string,
  name: string,
};

export interface ISymbolsState {
  data: ISymbol[] | null,
  error: Error | null,
  loading: boolean,
}

// Action types

export const LOAD_SYMBOLS_STARTED = `${DUCK_NAME}/LOAD_SYMBOLS_STARTED`;
interface ILoadSymbolsStarted {
  type: typeof LOAD_SYMBOLS_STARTED,
}

export const LOAD_SYMBOLS_SUCCEED = `${DUCK_NAME}/LOAD_SYMBOLS_SUCCEED`;
interface ILoadSymbolsSucceed {
  type: typeof LOAD_SYMBOLS_SUCCEED,
  data: ISymbol[],
}

export const LOAD_SYMBOLS_FAILED = `${DUCK_NAME}/LOAD_SYMBOLS_FAILED`;
interface ILoadSymbolsFailed {
  type: typeof LOAD_SYMBOLS_FAILED,
  error: Error,
}

// Action creators

export const loadSymbolsStarted = ():ILoadSymbolsStarted =>
  ({ type: LOAD_SYMBOLS_STARTED });

export const loadSymbolsSucceed = (data:ISymbol[]):ILoadSymbolsSucceed =>
  ({ type: LOAD_SYMBOLS_SUCCEED, data });

export const loadSymbolsFailed = (error:Error):ILoadSymbolsFailed =>
  ({ type: LOAD_SYMBOLS_FAILED, error });

// Thunks

export const loadSymbols = () =>
  async (dispatch:Dispatch) => {
    dispatch(loadSymbolsStarted());
    try {
      const results = await Api.instance.getSymbols();
      dispatch(loadSymbolsSucceed(results.data));
      return Promise.resolve(results);
    } catch (error) {
      dispatch(loadSymbolsFailed(error));
      return Promise.reject(error);
    }
  };

// Reducer

export const INITIAL_STATE:ISymbolsState = {
  data: null,
  error: null,
  loading: false,
};

const reducer = (state = INITIAL_STATE, action:AnyAction) => {
  switch (action.type) {
  case LOAD_SYMBOLS_STARTED:
    return {
      ...state,
      loading: true,
    };
  case LOAD_SYMBOLS_SUCCEED:
    return {
      ...state,
      data: [
        ...(state.data || []),
        ...action.data,
      ],
      loading: false,
    };
  case LOAD_SYMBOLS_FAILED:
    return {
      ...state,
      error: action.error,
      loading: false,
    };
  default: return state;
  }
};

export default reducer;
