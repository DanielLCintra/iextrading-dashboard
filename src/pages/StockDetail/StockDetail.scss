.StockDetail{
  &Header {
    text-align: center;
    margin-bottom: 30px;
  }

  &BackLink {
    text-decoration: none;
    display: inline;
  }
}
