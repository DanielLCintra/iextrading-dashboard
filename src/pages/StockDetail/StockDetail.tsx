import * as React from 'react';
import { RouteComponentProps } from 'react-router-dom';
import CompanyCard from '../../components/CompanyCard';
import Header from '../../components/Header';
import { ICompanyState } from '../../ducks/company';
import classes from './StockDetail.scss';

export interface IRouterProps {
  symbol: string,
};

export interface IStockDetailProps extends RouteComponentProps<IRouterProps> {
  company: ICompanyState,
  loadCompany: () => any,
}

class StockDetail extends React.Component<any, IStockDetailProps> {
  public componentDidMount() {
    const {
      loadCompany,
      match: {
        params: {
          symbol,
        },
      },
    } = this.props;
    loadCompany(symbol);
  }

  public render() {
    const {
      loading,
      error,
      data,
    } = this.props.company;

    if (loading || error || !data) {
      return null;
    }

    return (
      <div className={classes.wrapper}>
        <Header />
        <CompanyCard company={data} />
      </div>
    );
  }
}

export default StockDetail;
