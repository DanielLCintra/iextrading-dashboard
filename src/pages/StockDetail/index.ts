import { connect } from 'react-redux';
import { DUCK_NAME as COMPANY_DUCK_NAME, loadCompany } from '../../ducks/company';
import StockDetail, { IStockDetailProps } from './StockDetail';

const mapStateToProps = (
  state:any,
  ownProps:IStockDetailProps,
) => ({
  company: state[COMPANY_DUCK_NAME].get(ownProps.match.params.symbol) || {},
});

const mapDispatchToProps = (dispatch:any) => ({
  loadCompany: (symbol:string) => dispatch(loadCompany(symbol)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(StockDetail);
