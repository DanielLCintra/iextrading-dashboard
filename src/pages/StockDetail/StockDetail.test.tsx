import { shallow } from 'enzyme';
import { createBrowserHistory } from 'history';
import * as React from 'react';
import { INITIAL_COMPANY_STATE } from '../../ducks/company';
import StockDetail, { IStockDetailProps } from './StockDetail';

describe('StockDetail', () => {
  const symbol = 'TST';

  const mockLoadCompany = jest.fn();

  const mockMatchProp = {
    isExact: false,
    params: {
      symbol,
    },
    path: `/stock/${symbol}`,
    url: `http://test.com/stock/${symbol}`,
  };

  const props:IStockDetailProps = {
    company: INITIAL_COMPANY_STATE,
    history: createBrowserHistory(),
    loadCompany: mockLoadCompany,
    location: {
      hash: '',
      pathname: `/stock/${symbol}`,
      search: '',
      state: {},
    },
    match: mockMatchProp,
  };

  const wrapper = shallow(<StockDetail {...props} />);

  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should try to load the company data for the symbol', () => {
    expect(mockLoadCompany).toBeCalledWith(symbol);
  });

  it('should reload company data if symbol prop changes', () => {
    const newSymbol = 'CHANGE';

    mockLoadCompany.mockClear();

    wrapper.setProps({
      match: {
        ...mockMatchProp,
        params: {
          symbol: newSymbol
        },
      },
    });

    expect(mockLoadCompany).toBeCalledWith(newSymbol);
  });
});
