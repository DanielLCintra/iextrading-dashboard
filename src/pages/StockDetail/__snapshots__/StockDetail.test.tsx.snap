// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StockDetail renders without crashing 1`] = `""`;
